node/network/collator-protocol/src/collator_side.rs

```diff
@@ -658,59 +674,58 @@ async fn handle_our_view_change(
 	state: &mut State,
 	view: View,
 ) -> Result<()> {
-	let old_view = std::mem::replace(&mut (state.view), view);
-
-	let view = state.view.clone();
-
-	let removed = old_view.difference(&view).collect::<Vec<_>>();
-
-	for removed in removed.into_iter() {
+	for removed in state.view.difference(&view) {
 		state.collations.remove(removed);
 		state.our_validators_groups.remove(removed);
+		state.connection_requests.remove(removed);
 	}
 
+	state.view = view;
+
 	Ok(())
 }
 
 /// The collator protocol collator side main loop.
 #[tracing::instrument(skip(ctx, metrics), fields(subsystem = LOG_TARGET))]
-pub(crate) async fn run<Context>(
-	mut ctx: Context,
+pub(crate) async fn run(
+	mut ctx: impl SubsystemContext<Message = CollatorProtocolMessage>,
 	our_id: CollatorId,
 	metrics: Metrics,
-) -> Result<()>
-where
-	Context: SubsystemContext<Message = CollatorProtocolMessage>
-{
+) -> Result<()> {
 	use FromOverseer::*;
 	use OverseerSignal::*;
 
 	let mut state = State {
 		metrics,
+		our_id,
 		..Default::default()
 	};
 
-	state.our_id = our_id;
-
 	loop {
-		if let Some(mut request) = state.last_connection_request.take() {
-			let _timer = state.metrics.time_handle_connection_request();
-
-			while let Poll::Ready(Some((validator_id, peer_id))) = futures::poll!(request.next()) {
-				if let Err(err) = handle_validator_connected(&mut ctx, &mut state, peer_id, validator_id).await {
+		select! {
+			res = state.connection_requests.next().fuse() => {
+				let (relay_parent, validator_id, peer_id) = match res {
+					Some(res) => res,
+					None => continue,
```

Comment:
this is unreachable IIUC, it's fine, but maybe add a comment

Reply:
Yeah that is unreachable, another reason I had a `Future` before :D

Reply:
I would say we should write the caller code without relying on any quirkiness of the callee so that when after two years someone changes the behavior of the callee things do not break here. Usually saying something like "`malloc` does not fail anyway" is not ok. But this is not the hill I am willing to die on at the moment, so ok.

Reply:
I agree, ideally we want to enforce it via type system. But there is no abstraction for an infinite stream AFAIK, so emulating it with a stream that never returns an `Option` our best option so far.